**Indiquer dans la boîte ci-dessous vos noms et prénoms**

* Samy Oussada
* Nassim Messad

# 3I026 - Intelligence artificielle & Data science

# Projet

## Données

Vous devez travailler sur les données du projet Kaggle *European Soccer Database* accessible à l'adresse https://www.kaggle.com/hugomathien/soccer.

Les données en version SQLite sur la page http://www-connex.lip6.fr/~schwander/3I026/soccer.zip ou en version CSV http://www-connex.lip6.fr/~denoyer/wordpress/wp-content/uploads/2018/01/soccer.tar.gz. Pour facilter le travail dans les salles de la PPTI (et éviter de dépasser le quota de votre home), les données sont accessibles aussi dans `/Vrac/3I026/soccer/`.

Quel que soit le moyen pour récupérer le fichier de données, il est indispensable de lire en détail la page Kaggle pour comprendre à quoi ces données correspondent.

**Rappel:** il est attendu un notebook Python par problème, et le problème étudié doit être décrit précisément. Votre rapport devra impérativement suivre le format suivant.

In [1]:
import pandas as pd
import numpy as np

#  Préambule: analyse des données

### Lecture des tables 

In [2]:
country = pd.read_csv("soccer/country.csv")
country.shape

(11, 2)

In [3]:
league = pd.read_csv("soccer/league.csv")
league.shape

(11, 3)

In [4]:
player = pd.read_csv("soccer/player.csv")
player.shape

(11060, 7)

In [5]:
team = pd.read_csv("soccer/team.csv")
team.shape

(299, 5)

In [6]:
playerattributes = pd.read_csv("soccer/playerattributes.csv")
playerattributes.shape

(183978, 42)

In [7]:
teamattributes = pd.read_csv("soccer/teamattributes.csv")
teamattributes.shape

(1458, 25)

In [8]:
match = pd.read_csv("soccer/match.csv")
match.shape

(25979, 115)

 ### Etude des tables 

* Table Country

In [9]:
country

id         name
0       1      Belgium
1    1729      England
2    4769       France
3    7809      Germany
4   10257        Italy
5   13274  Netherlands
6   15722       Poland
7   17642     Portugal
8   19694     Scotland
9   21518        Spain
10  24558  Switzerland

In [10]:
country.shape

(11, 2)

La table country posséde deux colonnes et 11 tuples, country repressente des pays. <br> Les pays sont caractérisés par un "id" pour identifiant, et "name" pour le nom du pays.<br> les identifiants ne sont pas sequentiels, et les pays sont des pays europeens.

* Table league

In [11]:
league

id  country_id                      name
0       1           1    Belgium Jupiler League
1    1729        1729    England Premier League
2    4769        4769            France Ligue 1
3    7809        7809     Germany 1. Bundesliga
4   10257       10257             Italy Serie A
5   13274       13274    Netherlands Eredivisie
6   15722       15722        Poland Ekstraklasa
7   17642       17642  Portugal Liga ZON Sagres
8   19694       19694   Scotland Premier League
9   21518       21518           Spain LIGA BBVA
10  24558       24558  Switzerland Super League

In [12]:
league.shape

(11, 3)

In [13]:
(league['country_id']==country['id']).all()

True

La table league posséde 3 colonnes et 11 tuples. <br>
Chaque tuple represente une league, ceux si sont caractérisés par :
<ul>
    <li> id: represente l'id de la league
    <li> id_country: represente l'id du pays de la league
    <li> name: represente le nom de la league
</ul>
<i>id </i> et <i>id_country</i> sont equivalents

* Table team


In [14]:
team

id  team_api_id  team_fifa_api_id           team_long_name  \
0        1         9987             673.0                 KRC Genk   
1        2         9993             675.0             Beerschot AC   
2        3        10000           15005.0         SV Zulte-Waregem   
3        4         9994            2007.0         Sporting Lokeren   
4        5         9984            1750.0        KSV Cercle Brugge   
5        6         8635             229.0           RSC Anderlecht   
6        7         9991             674.0                 KAA Gent   
7        8         9998            1747.0                RAEC Mons   
8        9         7947               NaN            FCV Dender EH   
9       10         9985             232.0        Standard de Liège   
10      11         8203          110724.0              KV Mechelen   
11      12         8342             231.0           Club Brugge KV   
12      13         9999             546.0            KSV Roeselare   
13      14         8571          100081.0              KV Kortrijk   
14      15         4049               NaN                   Tubize   
15      16         9996          111560.0     Royal Excel Mouscron   
16      17        10001             681.0             KVC Westerlo   
17      18         9986             670.0       Sporting Charleroi   
18     614         9997             680.0        Sint-Truidense VV   
19    1034         9989             239.0                Lierse SK   
20    1042         6351            2013.0                KAS Eupen   
21    1513         1773          100087.0      Oud-Heverlee Leuven   
22    2004         8475          110913.0         Waasland-Beveren   
23    2476         8573             682.0              KV Oostende   
24    2510       274581          111560.0     Royal Excel Mouscron   
25    3457        10260              11.0        Manchester United   
26    3458        10261              13.0         Newcastle United   
27    3459         9825               1.0                  Arsenal   
28    3460         8659             109.0     West Bromwich Albion   
29    3461         8472             106.0               Sunderland   
..     ...          ...               ...                      ...   
269  43053         9906             240.0          Atlético Madrid   
270  43054         9864             573.0                Málaga CF   
271  43800         9868            1742.0     Xerez Club Deportivo   
272  43803         8394             244.0            Real Zaragoza   
273  43804         9867             260.0              CD Tenerife   
274  44557        10278          100879.0  Hércules Club de Fútbol   
275  44565         8581            1853.0               Levante UD   
276  44569         8560             457.0            Real Sociedad   
277  45330         7878          110832.0               Granada CF   
278  45333         8370             480.0           Rayo Vallecano   
279  46087         9910             450.0         RC Celta de Vigo   
280  46848        10268             468.0                 Elche CF   
281  47605         8372             467.0                 SD Eibar   
282  47612         7869            1867.0               Córdoba CF   
283  48358         8306             472.0            UD Las Palmas   
284  49115         9956             322.0  Grasshopper Club Zürich   
285  49116         6493            1714.0            AC Bellinzona   
286  49117        10192             900.0           BSC Young Boys   
287  49118         9931             896.0                 FC Basel   
288  49119         9930             434.0                 FC Aarau   
289  49120        10179          110770.0                  FC Sion   
290  49121        10199             897.0                FC Luzern   
291  49122         9824             286.0                 FC Vaduz   
292  49123         7955             435.0          Neuchâtel Xamax   
293  49124        10243             894.0                FC Zürich   
294  49479        

In [15]:
team.shape

(299, 5)

In [16]:
team.dtypes

id                    int64
team_api_id           int64
team_fifa_api_id    float64
team_long_name       object
team_short_name      object
dtype: object

In [17]:
id_=set(team.team_api_id)
print(len(id_))
id_fifa=set(team.team_fifa_api_id)
print(len(id_fifa))

299
296


La table team posséde 5 colonnes et 299 tuples. <br>
Chaque tuple represente une equipe, celles si sont caractérisées par :
<ul>
    <li> id: represente l'id de le ligne.
    <li> team_api_id: represente l'id de l'equipe.
    <li> team_fifa_api_id: represente l'id FIFA de l'equipe.
    <li> team_long_name: le nom de l'equipe.
    <li> team_short_name: le nom abrégé de l'equipe.
</ul>
<i>team_api_id </i> et <i>team_fifa_api_id</i> sont tous les deux des identifiants a valeur unique. <br>
<b> a noter qu'il y a des equipes qui n'ont pas d'identifiants FIFA 

* Table player


In [18]:
player

id  player_api_id             player_name  player_fifa_api_id  \
0          1         505942      Aaron Appindangoye              218353   
1          2         155782         Aaron Cresswell              189615   
2          3         162549             Aaron Doran              186170   
3          4          30572           Aaron Galindo              140161   
4          5          23780            Aaron Hughes               17725   
5          6          27316              Aaron Hunt              158138   
6          7         564793              Aaron Kuhl              221280   
7          8          30895            Aaron Lennon              152747   
8          9         528212            Aaron Lennox              206592   
9         10         101042           Aaron Meijers              188621   
10        11          23889           Aaron Mokoena               47189   
11        12         231592              Aaron Mooy              194958   
12        13         163222          Aaron Muirhead              213568   
13        14          40719            Aaron Niguez              183853   
14        15          75489            Aaron Ramsey              186561   
15        16         597948           Aaron Splaine              226014   
16        17         161644   Aaron Taylor-Sinclair              213569   
17        18          23499         Aaron Wilbraham                2335   
18        19         120919       Aatif Chahechouhe              187939   
19        20          46447               Abasse Ba              156626   
20        21         167027       Abdelaziz Barrada              192274   
21        22         245653   Abdelfettah Boukhriss              202425   
22        23         128456  Abdelhamid El Kaoutari              188145   
23        24          42664      Abdelkader Ghezzal              178063   
24        25         425950         Abdellah Zoubir              212934   
25        26          38423      Abdelmajid Oulmers               52782   
26        27           3264      Abdelmalek Cherrad               51868   
27        28         467485  Abdelmalek El Hasnaoui              209399   
28        29         306735     Abdelouahed Chakhsi              210504   
29        30          41659        Abderrazak Jadid              149241   
...      ...            ...                     ...                 ...   
11030  11046         244155                 Zezinho              212639   
11031  11047          19696               Zhi Zheng              158293   
11032  11048         290868             Zhi-Gin Lam              206209   
11033  11049         169594             Zie Diabate              207780   
11034  11050         244953            Ziggy Gordon              223263   
11035  11051         200607         Ziguy Badibanga              195862   
11036  11052         642618        Zinedine Machach              228821   
11037  11053         277772              Zinho Gano              204227   
11038  11054         246332             Ziri Hammar              202751   
11039  11055         532766                    Zizo              222537   
11040  11056          27305       Zlatan Bajramovic               31667   
11041  11057          35724      Zlatan Ibrahimovic               41236   
11042  11058          12574      Zlatan Ljubijankic              176103   
11043  11059          39494            Zlatko Dedic              141111   
11044  11060          17233           Zlatko Janjic              188405   
11045  11061          42860        Zlatko Junuzovic              173474   
11046  11062          37128           Zola Matumona              158142   
11047  11063          24737             Zoltan Gera              137089   
11048  11064         107281          Zoltan Stieber              179712   
11049  11065          39801          Zoltan Szelesi              157283   
11050  11066         491794         Zoran Josipovic              219943   
11051  11067          56929        

In [19]:
player.dtypes

id                      int64
player_api_id           int64
player_name            object
player_fifa_api_id      int64
birthday               object
height                float64
weight                  int64
dtype: object

In [20]:
id_=set(player.player_api_id)
print(len(id_))
id_fifa=set(player.player_fifa_api_id)
print(len(id_fifa))

11060
11060


La table player posséde 7 colonnes et 11060 tuples. <br>
Chaque tuple represente un joueur, celles si sont caractérisées par :
<ul>
    <li> id: represente l'id de le ligne.
    <li> player_api_id: represente l'id du joueur.
    <li> player_fifa_api_id: represente l'id FIFA du joueur.
    <li> birthday: la date de naissance du joueur.
    <li> height: la taille du joueur en centimetres.
    <li> weight: le poids du joueur en livres
</ul>
<i>player_api_id </i> et <i>player_fifa_api_id</i> sont tous les deux des identifiants a valeur unique. <br>


* Table playerattributes


In [21]:
playerattributes

id  player_fifa_api_id  player_api_id                 date  \
0            1              218353         505942  2016-02-18 00:00:00   
1            2              218353         505942  2015-11-19 00:00:00   
2            3              218353         505942  2015-09-21 00:00:00   
3            4              218353         505942  2015-03-20 00:00:00   
4            5              218353         505942  2007-02-22 00:00:00   
5            6              189615         155782  2016-04-21 00:00:00   
6            7              189615         155782  2016-04-07 00:00:00   
7            8              189615         155782  2016-01-07 00:00:00   
8            9              189615         155782  2015-12-24 00:00:00   
9           10              189615         155782  2015-12-17 00:00:00   
10          11              189615         155782  2015-10-16 00:00:00   
11          12              189615         155782  2015-09-25 00:00:00   
12          13              189615         155782  2015-09-21 00:00:00   
13          14              189615         155782  2015-01-09 00:00:00   
14          15              189615         155782  2014-12-05 00:00:00   
15          16              189615         155782  2014-11-07 00:00:00   
16          17              189615         155782  2014-09-18 00:00:00   
17          18              189615         155782  2014-05-02 00:00:00   
18          19              189615         155782  2014-04-04 00:00:00   
19          20              189615         155782  2014-03-14 00:00:00   
20          21              189615         155782  2013-12-13 00:00:00   
21          22              189615         155782  2013-11-08 00:00:00   
22          23              189615         155782  2013-10-04 00:00:00   
23          24              189615         155782  2013-09-20 00:00:00   
24          25              189615         155782  2013-05-03 00:00:00   
25          26              189615         155782  2013-03-22 00:00:00   
26          27              189615         155782  2013-03-15 00:00:00   
27          28              189615         155782  2013-02-22 00:00:00   
28          29              189615         155782  2013-02-15 00:00:00   
29          30              189615         155782  2012-08-31 00:00:00   
...        ...                 ...            ...                  ...   
183948  183949              164680         111182  2012-02-22 00:00:00   
183949  183950              164680         111182  2011-08-30 00:00:00   
183950  183951              164680         111182  2011-02-22 00:00:00   
183951  183952              164680         111182  2010-08-30 00:00:00   
183952  183953              164680         111182  2007-02-22 00:00:00   
183953  183954              111191          36491  2011-02-22 00:00:00   
183954  183955              111191          36491  2010-08-30 00:00:00   
183955  183956              111191          36491  2009-08-30 00:00:00   
183956  183957              111191          36491  2009-02-22 00:00:00   
183957  183958              111191          36491  2008-08-30 00:00:00   
183958  183959              111191          36491  2007-08-30 00:00:00   
183959  183960              111191          36491  2007-02-22 00:00:00   
183960  183961               47058          35506  2011-02-22 00:00:00   
183961  183962               47058          35506  2010-08-30 00:00:00   
183962  183963               47058          35506  2010-02-22 00:00:00   
183963  183964               47058          35506  2009-08-30 00:00:00   
183964  183965               47058          35506  2009-02-22 00:00:00   
183965  183966               47058          35506  2008-08-30 00:00:00   
183966  183967               47058          35506  2007-08-30 00:00:00   
183967  183968               47058          35506  2007-02-22 00:00:00   
183968  183969              102359          39902  2012-08-31 00:00:00   
183969  183970              102359          39902  2012-02-22 00:00:00   
183970  183971  

In [22]:

def isNaN(x):
    return str((x)).lower() == 'nan'
for col in playerattributes.columns:
    for i in range(playerattributes.shape[0]):
        if (isNaN(playerattributes[col][i])):
            print("la colonne: "+str(col)+" posséde des tuples non renseignés")
            break
            
print("fininini")

la colonne: overall_rating posséde des tuples non renseignés
la colonne: potential posséde des tuples non renseignés
la colonne: preferred_foot posséde des tuples non renseignés
la colonne: attacking_work_rate posséde des tuples non renseignés
la colonne: defensive_work_rate posséde des tuples non renseignés
la colonne: crossing posséde des tuples non renseignés
la colonne: finishing posséde des tuples non renseignés
la colonne: heading_accuracy posséde des tuples non renseignés
la colonne: short_passing posséde des tuples non renseignés
la colonne: volleys posséde des tuples non renseignés
la colonne: dribbling posséde des tuples non renseignés
la colonne: curve posséde des tuples non renseignés
la colonne: free_kick_accuracy posséde des tuples non renseignés
la colonne: long_passing posséde des tuples non renseignés
la colonne: ball_control posséde des tuples non renseignés
la colonne: acceleration posséde des tuples non renseignés
la colonne: sprint_speed posséde des tuples non rens

In [23]:
playerattributes.dtypes

id                       int64
player_fifa_api_id       int64
player_api_id            int64
date                    object
overall_rating         float64
potential              float64
preferred_foot          object
attacking_work_rate     object
defensive_work_rate     object
crossing               float64
finishing              float64
heading_accuracy       float64
short_passing          float64
volleys                float64
dribbling              float64
curve                  float64
free_kick_accuracy     float64
long_passing           float64
ball_control           float64
acceleration           float64
sprint_speed           float64
agility                float64
reactions              float64
balance                float64
shot_power             float64
jumping                float64
stamina                float64
strength               float64
long_shots             float64
aggression             float64
interceptions          float64
positioning            float64
vision  

La table playerattributes posséde 42 colonnes et 183978 tuples. <br>
Chaque tuple represente un joueur, ceux si sont caractérisés par :
<ul>
    <li> id: represente l'id de la ligne.
    <li> player_fifa_api_id: represente l'id FIFA du joueur.
    <li> player_api_id: represente l'id du joueur.
    <li> date: represente la date correspondante aux informations relatives au joueur.
    <li> overall_rating: la note generale du joueur.
    <li> potential: le potentiel du joueur
    <li> preferred_foot: le pied fort du joueur
    <li> attacking_work_rate: note global sur les capacités offensives du joueur
    <li> defensive_work_rate: note global sur les capacités defensives du joueur
    <li> Le reste des colonnes representent les notes detaillés sur des aspects specifiques : techniques et athletiques
</ul>
<i>player_api_id </i> et <i>player_fifa_api_id</i> et <i>date</i> sont utilisés comme identifiant de chaque tuple. <br>


* Table teamattributes

In [24]:
teamattributes

id  team_fifa_api_id  team_api_id                 date  \
0        1               434         9930  2010-02-22 00:00:00   
1        2               434         9930  2014-09-19 00:00:00   
2        3               434         9930  2015-09-10 00:00:00   
3        4                77         8485  2010-02-22 00:00:00   
4        5                77         8485  2011-02-22 00:00:00   
5        6                77         8485  2012-02-22 00:00:00   
6        7                77         8485  2013-09-20 00:00:00   
7        8                77         8485  2014-09-19 00:00:00   
8        9                77         8485  2015-09-10 00:00:00   
9       10               614         8576  2010-02-22 00:00:00   
10      11               614         8576  2011-02-22 00:00:00   
11      12               614         8576  2012-02-22 00:00:00   
12      13               614         8576  2013-09-20 00:00:00   
13      14               614         8576  2014-09-19 00:00:00   
14      15               614         8576  2015-09-10 00:00:00   
15      16                47         8564  2010-02-22 00:00:00   
16      17                47         8564  2011-02-22 00:00:00   
17      18                47         8564  2012-02-22 00:00:00   
18      19                47         8564  2013-09-20 00:00:00   
19      20                47         8564  2014-09-19 00:00:00   
20      21                47         8564  2015-09-10 00:00:00   
21      22              1901        10215  2010-02-22 00:00:00   
22      23              1901        10215  2011-02-22 00:00:00   
23      24              1901        10215  2012-02-22 00:00:00   
24      25              1901        10215  2013-09-20 00:00:00   
25      26              1901        10215  2014-09-19 00:00:00   
26      27              1901        10215  2015-09-10 00:00:00   
27      28               650        10217  2010-02-22 00:00:00   
28      29               650        10217  2011-02-22 00:00:00   
29      30               650        10217  2012-02-22 00:00:00   
...    ...               ...          ...                  ...   
1428  1429               435         7955  2012-02-22 00:00:00   
1429  1430              1742         9868  2010-02-22 00:00:00   
1430  1431              1742         9868  2011-02-22 00:00:00   
1431  1432              1742         9868  2012-02-22 00:00:00   
1432  1433              1742         9868  2013-09-20 00:00:00   
1433  1434               900        10192  2010-02-22 00:00:00   
1434  1435               900        10192  2011-02-22 00:00:00   
1435  1436               900        10192  2012-02-22 00:00:00   
1436  1437               900        10192  2013-09-20 00:00:00   
1437  1438               900        10192  2014-09-19 00:00:00   
1438  1439               900        10192  2015-09-10 00:00:00   
1439  1440            110749         8021  2010-02-22 00:00:00   
1440  1441            110749         8021  2011-02-22 00:00:00   
1441  1442            110749         8021  2012-02-22 00:00:00   
1442  1443            110749         8021  2013-09-20 00:00:00   
1443  1444            110749         8021  2014-09-19 00:00:00   
1444  1445               244         8394  2010-02-22 00:00:00   
1445  1446               244         8394  2011-02-22 00:00:00   
1446  1447               244         8394  2012-02-22 00:00:00   
1447  1448               244         8394  2013-09-20 00:00:00   
1448  1449               244         8394  2014-09-19 00:00:00   
1449  1450               244         8394  2015-09-10 00:00:00   
1450  1451            112512         8027  2014-09-19 00:00:00   
1451  1452            112512         8027  2015-09-10 00:00:00   
1452  1453             15005        10000  2010-02-22 00:00:00   
1453  1454             15005        10000  2011-02-22 00:00:00   
1454  1455             15005        10000  2012-02-22 00:00:00   
1455  1456             15005        10000  2013-09-20 00:00:00   
1456  1457             15005        10000  2014-

In [25]:

def isNaN(x):
    return str((x)).lower() == 'nan'

for col in teamattributes.columns:
    for i in range(teamattributes.shape[0]):
        if (isNaN(teamattributes[col][i])):
            print("la colonne: "+str(col)+" posséde des tuples non renseignés")
            break
            
print("fin")

la colonne: buildUpPlayDribbling posséde des tuples non renseignés
fin


In [26]:
teamattributes.dtypes

id                                  int64
team_fifa_api_id                    int64
team_api_id                         int64
date                               object
buildUpPlaySpeed                    int64
buildUpPlaySpeedClass              object
buildUpPlayDribbling              float64
buildUpPlayDribblingClass          object
buildUpPlayPassing                  int64
buildUpPlayPassingClass            object
buildUpPlayPositioningClass        object
chanceCreationPassing               int64
chanceCreationPassingClass         object
chanceCreationCrossing              int64
chanceCreationCrossingClass        object
chanceCreationShooting              int64
chanceCreationShootingClass        object
chanceCreationPositioningClass     object
defencePressure                     int64
defencePressureClass               object
defenceAggression                   int64
defenceAggressionClass             object
defenceTeamWidth                    int64
defenceTeamWidthClass             

La table teamattributes posséde 25 colonnes et 1458 tuples. <br>
Chaque tuple represente une equipe, celles si sont caractérisées par :
<ul>
    <li> id: represente l'id de la ligne.
    <li> team_fifa_api_id: represente l'id FIFA de l'equipe.
    <li> team_api_id: represente l'id dde l'equipe.
    <li> date: represente la date correspondante aux informations relatives a l'equipe.
    <li> Le reste des colonnes represente des notes sur l'aspect du jeu de l'equipe suivi de la classe de cette aspect
</ul>
<i>team_api_id </i> et <i>team_fifa_api_id</i> et <i>date</i> sont utilisés comme identifiant de chaque tuple. <br>


* Table match


In [27]:

def isNaN(x):
    return str((x)).lower() == 'nan'

for col in match.columns:
    for i in range(match.shape[0]):
        if (isNaN(match[col][i])):
            print("la colonne: "+str(col)+" posséde des tuples non renseignés")
            break
            
print("fin")

la colonne: home_player_X1 posséde des tuples non renseignés
la colonne: home_player_X2 posséde des tuples non renseignés
la colonne: home_player_X3 posséde des tuples non renseignés
la colonne: home_player_X4 posséde des tuples non renseignés
la colonne: home_player_X5 posséde des tuples non renseignés
la colonne: home_player_X6 posséde des tuples non renseignés
la colonne: home_player_X7 posséde des tuples non renseignés
la colonne: home_player_X8 posséde des tuples non renseignés
la colonne: home_player_X9 posséde des tuples non renseignés
la colonne: home_player_X10 posséde des tuples non renseignés
la colonne: home_player_X11 posséde des tuples non renseignés
la colonne: away_player_X1 posséde des tuples non renseignés
la colonne: away_player_X2 posséde des tuples non renseignés
la colonne: away_player_X3 posséde des tuples non renseignés
la colonne: away_player_X4 posséde des tuples non renseignés
la colonne: away_player_X5 posséde des tuples non renseignés
la colonne: away_playe

In [28]:
colonnes=match.columns
match

id  country_id  league_id     season  stage                 date  \
0          1           1          1  2008/2009      1  2008-08-17 00:00:00   
1          2           1          1  2008/2009      1  2008-08-16 00:00:00   
2          3           1          1  2008/2009      1  2008-08-16 00:00:00   
3          4           1          1  2008/2009      1  2008-08-17 00:00:00   
4          5           1          1  2008/2009      1  2008-08-16 00:00:00   
5          6           1          1  2008/2009      1  2008-09-24 00:00:00   
6          7           1          1  2008/2009      1  2008-08-16 00:00:00   
7          8           1          1  2008/2009      1  2008-08-16 00:00:00   
8          9           1          1  2008/2009      1  2008-08-16 00:00:00   
9         10           1          1  2008/2009     10  2008-11-01 00:00:00   
10        11           1          1  2008/2009     10  2008-10-31 00:00:00   
11        12           1          1  2008/2009     10  2008-11-02 00:00:00   
12        13           1          1  2008/2009     10  2008-11-01 00:00:00   
13        14           1          1  2008/2009     10  2008-11-01 00:00:00   
14        15           1          1  2008/2009     10  2008-11-01 00:00:00   
15        16           1          1  2008/2009     10  2008-11-01 00:00:00   
16        17           1          1  2008/2009     10  2008-11-01 00:00:00   
17        18           1          1  2008/2009     10  2008-11-02 00:00:00   
18        19           1          1  2008/2009     11  2008-11-08 00:00:00   
19        20           1          1  2008/2009     11  2008-11-08 00:00:00   
20        21           1          1  2008/2009     11  2008-11-09 00:00:00   
21        22           1          1  2008/2009     11  2008-11-07 00:00:00   
22        23           1          1  2008/2009     11  2008-11-08 00:00:00   
23        24           1          1  2008/2009     11  2008-11-08 00:00:00   
24        25           1          1  2008/2009     11  2008-11-08 00:00:00   
25        26           1          1  2008/2009     11  2008-11-08 00:00:00   
26        27           1          1  2008/2009     11  2008-11-09 00:00:00   
27        28           1          1  2008/2009     12  2008-11-16 00:00:00   
28        29           1          1  2008/2009     12  2008-11-15 00:00:00   
29        30           1          1  2008/2009     12  2008-11-15 00:00:00   
...      ...         ...        ...        ...    ...                  ...   
25949  25950       24558      24558  2015/2016      4  2015-08-08 00:00:00   
25950  25951       24558      24558  2015/2016      4  2015-08-08 00:00:00   
25951  25952       24558      24558  2015/2016      4  2015-08-09 00:00:00   
25952  25953       24558      24558  2015/2016      4  2015-08-09 00:00:00   
25953  25954       24558      24558  2015/2016      4  2015-08-09 00:00:00   
25954  25955       24558      24558  2015/2016      5  2015-08-12 00:00:00   
25955  25956       24558      24558  2015/2016      5  2015-08-12 00:00:00   
25956  25957       24558      24558  2015/2016      5  2015-08-12 00:00:00   
25957  25958       24558      24558  2015/2016      5  2015-08-13 00:00:00   
25958  25959       24558      24558  2015/2016      5  2015-08-13 00:00:00   
25959  25960       24558      24558  2015/2016      6  2015-08-22 00:00:00   
25960  25961       24558      24558  2015/2016      6  2015-08-23 00:00:00   
25961  25962       24558      24558  2015/2016      6  2015-08-23 00:00:00   
25962  25963       24558      24558  2015/2016      6  2015-08-23 00:00:00   
25963  25964       24558      24558  2015/2016      6  2015-08-22 00:00:00   
25964  25965       24558      24558  2015/2016      7  2015-08-29 00:00:00   
25965  25966       24558      24558  2015/2016      7  2015-08-29 00:00:00   
25966  25967       24558      24558  2015/2016      7  2015-08-30 00:00:00   
25967  25968       24558      24558  2015/2016      7  2015-08-30 00:00:00   
25968  25969       24558      24

In [29]:
match.dtypes[0:60]

id                    int64
country_id            int64
league_id             int64
season               object
stage                 int64
date                 object
match_api_id          int64
home_team_api_id      int64
away_team_api_id      int64
home_team_goal        int64
away_team_goal        int64
home_player_X1      float64
home_player_X2      float64
home_player_X3      float64
home_player_X4      float64
home_player_X5      float64
home_player_X6      float64
home_player_X7      float64
home_player_X8      float64
home_player_X9      float64
home_player_X10     float64
home_player_X11     float64
away_player_X1      float64
away_player_X2      float64
away_player_X3      float64
away_player_X4      float64
away_player_X5      float64
away_player_X6      float64
away_player_X7      float64
away_player_X8      float64
away_player_X9      float64
away_player_X10     float64
away_player_X11     float64
home_player_Y1      float64
home_player_Y2      float64
home_player_Y3      

In [30]:
match.dtypes[60:]

home_player_6     float64
home_player_7     float64
home_player_8     float64
home_player_9     float64
home_player_10    float64
home_player_11    float64
away_player_1     float64
away_player_2     float64
away_player_3     float64
away_player_4     float64
away_player_5     float64
away_player_6     float64
away_player_7     float64
away_player_8     float64
away_player_9     float64
away_player_10    float64
away_player_11    float64
goal               object
shoton             object
shotoff            object
foulcommit         object
card               object
cross              object
corner             object
possession         object
B365H             float64
B365D             float64
B365A             float64
BWH               float64
BWD               float64
BWA               float64
IWH               float64
IWD               float64
IWA               float64
LBH               float64
LBD               float64
LBA               float64
PSH               float64
PSD         

In [31]:

l = [i for i in list(match["home_player_X1"]) if str((i)).lower() != 'nan']
set(l)

{0.0, 1.0, 2.0}

In [32]:

l = [i for i in list(match["home_player_X2"]) if str((i)).lower() != 'nan']
set(l)

{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0}

In [33]:

l = [i for i in list(match["home_player_X11"]) if str((i)).lower() != 'nan']
set(l)

{1.0, 3.0, 4.0, 5.0, 6.0, 7.0}

In [34]:

l = [i for i in list(match["home_player_Y1"]) if str((i)).lower() != 'nan']
set(l)

{0.0, 1.0, 3.0}

In [35]:

l = [i for i in list(match["home_player_Y11"]) if str((i)).lower() != 'nan']
set(l)

{1.0, 3.0, 10.0, 11.0}

In [36]:

#l = [i for i in list(match["home_player_8"]) if str((i)).lower() != 'nan']
#set(l)

In [37]:
print(len(match))
nunu=match.isnull().sum()
nunu[50:]

25979


away_player_Y7      1832
away_player_Y8      1832
away_player_Y9      1833
away_player_Y10     1833
away_player_Y11     1839
home_player_1       1224
home_player_2       1315
home_player_3       1281
home_player_4       1323
home_player_5       1316
home_player_6       1325
home_player_7       1227
home_player_8       1309
home_player_9       1273
home_player_10      1436
home_player_11      1555
away_player_1       1234
away_player_2       1278
away_player_3       1293
away_player_4       1321
away_player_5       1335
away_player_6       1313
away_player_7       1235
away_player_8       1341
away_player_9       1328
away_player_10      1441
away_player_11      1554
goal               11762
shoton             11762
shotoff            11762
                   ...  
B365H               3387
B365D               3387
B365A               3387
BWH                 3404
BWD                 3404
BWA                 3404
IWH                 3459
IWD                 3459
IWA                 3459


In [38]:
ideq=team['team_api_id'].iloc[20]
for i in range(len(match)):
    if (match["home_team_api_id"].iloc[i] == ideq) or(match["away_team_api_id"].iloc[i]):
        print(match["home_player_1"].iloc[i])

nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
39890.0
38327.0
95597.0
nan
30934.0
37990.0
38391.0
39153.0
37900.0
36835.0
38252.0
34480.0
33676.0
37937.0
38341.0
38318.0
104378.0
95597.0
38327.0
nan
38391.0
30934.0
39153.0
37990.0
37900.0
148308.0
32990.0
33676.0
95597.0
104378.0
38252.0
38341.0
38318.0
36835.0
37937.0
34480.0
30934.0
148308.0
37990.0
37900.0
39153.0
32990.0
39573.0
37937.0
38797.0
34480.0
38318.0
38391.0
38797.0
33676.0
38252.0
95597.0
36835.0
nan


24527.0
30455.0
30380.0
22978.0
31432.0
36286.0
23794.0
34421.0
30859.0
30660.0
32562.0
30455.0
30380.0
22978.0
31432.0
36286.0
23794.0
34421.0
30859.0
30726.0
23686.0
36374.0
31465.0
24224.0
30622.0
30633.0
23932.0
36373.0
23021.0
30660.0
24527.0
30455.0
30380.0
22978.0
31432.0
36286.0
23794.0
34421.0
30859.0
30726.0
23686.0
36374.0
31465.0
24224.0
30622.0
30633.0
23932.0
36373.0
23021.0
30660.0
24527.0
30455.0
30380.0
22978.0
31432.0
36286.0
23794.0
34421.0
30859.0
30660.0
30455.0
24224.0
31432.0
30622.0
30633.0
36373.0
36286.0
23794.0
24886.0
30726.0
23686.0
36374.0
31465.0
24224.0
30622.0
30633.0
23932.0
36373.0
23021.0
30726.0
23686.0
30380.0
31465.0
22978.0
24224.0
30633.0
36286.0
23794.0
34421.0
30660.0
24527.0
30455.0
36374.0
31432.0
30622.0
23932.0
36373.0
23021.0
30859.0
24527.0
23984.0
36374.0
36373.0
36286.0
24886.0
31465.0
31432.0
30622.0
30859.0
30726.0
23686.0
30660.0
30380.0
35442.0
24227.0
30633.0
23932.0
23027.0
23794.0
24527.0
23984.0
36374.0
31465.0
24224.0
30622.0


35477.0
182917.0
24579.0
30989.0
38617.0
46518.0
37868.0
22978.0
30660.0
31432.0
24229.0
30974.0
37193.0
26295.0
24147.0
23932.0
25156.0
169718.0
69650.0
31465.0
30633.0
182917.0
24579.0
30989.0
38617.0
46518.0
37868.0
30859.0
35442.0
31432.0
24229.0
30974.0
37193.0
26295.0
24147.0
23932.0
25156.0
169718.0
69650.0
182917.0
24579.0
30989.0
38617.0
46518.0
37868.0
31465.0
30633.0
169718.0
31465.0
30633.0
31432.0
24229.0
30989.0
38617.0
24147.0
23932.0
25156.0
69650.0
30859.0
30660.0
182917.0
24579.0
30974.0
46518.0
37868.0
37193.0
26295.0
169718.0
30633.0
30660.0
31432.0
23979.0
24579.0
30989.0
38617.0
24147.0
23932.0
69650.0
30859.0
24341.0
182917.0
30974.0
46518.0
37868.0
37193.0
26295.0
25156.0
169718.0
30633.0
35442.0
31432.0
23979.0
24579.0
36374.0
23793.0
24147.0
23932.0
30859.0
30660.0
31432.0
24229.0
35496.0
37193.0
30380.0
24147.0
23932.0
25156.0
69650.0
30859.0
24341.0
182917.0
30974.0
46518.0
37868.0
37193.0
26295.0
186054.0
30973.0
69650.0
31465.0
30633.0
31432.0
30974.0
3786

24147.0
50065.0
69650.0
30974.0
37770.0
37868.0
182917.0
40602.0
188971.0
37233.0
169756.0
69650.0
36283.0
37868.0
31432.0
95336.0
37233.0
30973.0
26295.0
30455.0
169756.0
30859.0
30974.0
170323.0
186054.0
37770.0
182917.0
163604.0
40602.0
24224.0
24147.0
69650.0
30974.0
36283.0
37770.0
37868.0
31432.0
40602.0
37233.0
26295.0
37780.0
30859.0
46518.0
31465.0
182917.0
163604.0
95336.0
188971.0
30973.0
30455.0
24147.0
24579.0
30974.0
36283.0
37770.0
37868.0
31432.0
40602.0
37233.0
26295.0
37780.0
30859.0
170323.0
31465.0
182917.0
163604.0
95336.0
24224.0
73462.0
30455.0
24147.0
69650.0
30974.0
170323.0
37868.0
31432.0
24229.0
30841.0
288880.0
30973.0
26295.0
30859.0
119079.0
31465.0
37770.0
182917.0
40602.0
103428.0
30455.0
23021.0
37780.0
69650.0
30974.0
46518.0
37868.0
31432.0
24229.0
30841.0
288880.0
30973.0
26295.0
37770.0
37868.0
182917.0
24229.0
30841.0
288880.0
26295.0
30455.0
23021.0
169756.0
30859.0
69650.0
30974.0
46518.0
36283.0
31465.0
31432.0
40602.0
103428.0
30973.0
46518.0


31293.0
31047.0
26168.0
26275.0
95597.0
30742.0
40709.0
26167.0
50065.0
159883.0
41301.0
31293.0
46705.0
78904.0
27659.0
30900.0
11319.0
41097.0
26335.0
190741.0
159883.0
31047.0
41886.0
26168.0
26275.0
95597.0
11320.0
40709.0
26167.0
50065.0
41301.0
46705.0
26290.0
50065.0
30900.0
11319.0
41097.0
26335.0
190741.0
159883.0
27659.0
31047.0
41886.0
26168.0
26275.0
11320.0
40709.0
26167.0
31293.0
190741.0
41301.0
31047.0
140951.0
26168.0
26275.0
30742.0
40709.0
26295.0
31293.0
190741.0
41301.0
95597.0
33109.0
26290.0
94306.0
27659.0
30900.0
11319.0
41097.0
26335.0
159883.0
31047.0
30742.0
26168.0
26275.0
11320.0
40709.0
26167.0
31293.0
26335.0
190741.0
95597.0
46705.0
26290.0
50065.0
27659.0
103468.0
11319.0
41097.0
159883.0
41301.0
31047.0
30742.0
26168.0
26275.0
11320.0
210114.0
31293.0
46705.0
26335.0
190741.0
31047.0
95597.0
26167.0
26290.0
50065.0
27659.0
11319.0
41097.0
159883.0
41301.0
30742.0
26168.0
26275.0
11320.0
40709.0
31293.0
46705.0
30900.0
26335.0
190741.0
31047.0
26168.0


30820.0
nan
36058.0
27255.0
33339.0
38793.0
37377.0
25524.0
27358.0
27313.0
24104.0
27467.0
26173.0
31299.0
27341.0
42909.0
27284.0
27424.0
30648.0
49586.0
30820.0
36058.0
33339.0
36029.0
42909.0
25524.0
27358.0
27313.0
24104.0
27467.0
26173.0
31299.0
27341.0
37377.0
27284.0
27424.0
27358.0
30648.0
27299.0
36058.0
33339.0
36029.0
29303.0
25524.0
27313.0
24104.0
30820.0
27467.0
26173.0
31299.0
27518.0
37377.0
27284.0
27424.0
27358.0
30648.0
27299.0
33339.0
27518.0
nan
29303.0
30820.0
27358.0
27313.0
24104.0
36343.0
27433.0
38668.0
25018.0
30703.0
32653.0
27424.0
24104.0
27299.0
38668.0
27518.0
25018.0
38793.0
30703.0
25524.0
27358.0
30648.0
27299.0
30820.0
26335.0
27467.0
26173.0
27433.0
32653.0
114963.0
27313.0
24104.0
38668.0
27341.0
25018.0
38793.0
30703.0
32653.0
25524.0
27358.0
30648.0
27299.0
30820.0
26335.0
27467.0
26173.0
27424.0
27313.0
24104.0
71524.0
38668.0
27518.0
25018.0
38793.0
30703.0
32653.0
25524.0
27358.0
30648.0
27299.0
30820.0
27467.0
27255.0
27518.0
27313.0
24104.0

184554.0
94841.0
94139.0
27433.0
178778.0
27424.0
177698.0
145550.0
27299.0
49586.0
215168.0
41087.0
30820.0
97746.0
287894.0
114737.0
36878.0
178778.0
30703.0
184554.0
94841.0
41087.0
30820.0
27433.0
27424.0
145550.0
27358.0
27299.0
215168.0
128689.0
170322.0
25018.0
39566.0
5440.0
36878.0
30703.0
184554.0
94841.0
41087.0
30820.0
27433.0
178778.0
25018.0
27424.0
27358.0
27299.0
215168.0
128689.0
170322.0
39566.0
5440.0
145550.0
36878.0
30703.0
184554.0
215168.0
36479.0
41087.0
30820.0
27433.0
178778.0
25018.0
27358.0
27299.0
128689.0
170322.0
39566.0
27424.0
177698.0
145550.0
36878.0
27299.0
69149.0
41618.0
178778.0
41087.0
36878.0
145550.0
212815.0
37335.0
27299.0
49586.0
41618.0
287894.0
41087.0
36479.0
24104.0
212815.0
37335.0
27358.0
215168.0
30820.0
114737.0
178778.0
36878.0
94139.0
177953.0
112978.0
27299.0
49586.0
41618.0
287894.0
41087.0
94841.0
24104.0
212815.0
37335.0
49586.0
215168.0
30820.0
39566.0
178778.0
36878.0
94139.0
177953.0
112978.0
27299.0
69149.0
41618.0
41087.0


41243.0
27560.0
42502.0
41643.0
39477.0
42422.0
34530.0
27691.0
27659.0
41671.0
35644.0
30974.0
41818.0
39401.0
32746.0
41671.0
30717.0
43283.0
30989.0
39415.0
41643.0
56829.0
27560.0
42502.0
34530.0
35626.0
39401.0
39477.0
27691.0
41818.0
41243.0
30974.0
27659.0
35644.0
42502.0
35626.0
35644.0
30989.0
41671.0
56829.0
43283.0
27560.0
32746.0
41643.0
41243.0
30717.0
27659.0
34530.0
39401.0
39415.0
41643.0
27691.0
41818.0
42502.0
42422.0
30989.0
39769.0
32746.0
30974.0
43283.0
41671.0
39477.0
27560.0
35644.0
39415.0
39351.0
41643.0
27691.0
41243.0
34530.0
41818.0
27615.0
27659.0
42502.0
27560.0
39477.0
30989.0
30717.0
56829.0
27691.0
30974.0
41671.0
32746.0
42422.0
42502.0
35644.0
39415.0
41243.0
27560.0
39351.0
39477.0
34530.0
43283.0
41643.0
42422.0
27691.0
39401.0
35644.0
27659.0
34530.0
24503.0
39477.0
41243.0
41818.0
32746.0
30974.0
27691.0
30717.0
27659.0
41671.0
30989.0
42422.0
41818.0
56829.0
32746.0
39415.0
27560.0
39351.0
34530.0
39477.0
41243.0
43283.0
41643.0
35644.0
42422.0


157284.0
27558.0
41045.0
27615.0
163255.0
37503.0
39599.0
35644.0
41466.0
13033.0
27658.0
181910.0
39415.0
42692.0
105828.0
212630.0
42422.0
41671.0
39170.0
157284.0
176186.0
32746.0
37503.0
27658.0
39477.0
35644.0
181910.0
30717.0
39599.0
39294.0
13033.0
34530.0
394824.0
103218.0
42422.0
74028.0
105828.0
212630.0
178732.0
41671.0
157284.0
282706.0
163255.0
176186.0
553849.0
42422.0
30717.0
39599.0
172157.0
35644.0
41671.0
212630.0
213299.0
181920.0
181910.0
42422.0
39477.0
32746.0
39170.0
41818.0
13033.0
163255.0
176186.0
105828.0
30717.0
39599.0
34530.0
172157.0
35644.0
41671.0
13033.0
212630.0
282706.0
176186.0
181910.0
105828.0
42422.0
39477.0
32746.0
39170.0
41818.0
213299.0
181920.0
39294.0
30717.0
39599.0
34530.0
172157.0
35644.0
109060.0
13033.0
212630.0
282706.0
176186.0
181910.0
105828.0
42422.0
39477.0
178732.0
39170.0
41818.0
178732.0
103218.0
213299.0
105828.0
42422.0
30717.0
150747.0
39477.0
41671.0
39170.0
212630.0
155246.0
213299.0
553849.0
618878.0
30660.0
169718.0
416

118848.0
37216.0
26547.0
150236.0
37206.0
26002.0
42829.0
73462.0
45836.0
214984.0
26484.0
26536.0
5710.0
36846.0
206809.0
42592.0
118848.0
37216.0
112107.0
21645.0
214984.0
73462.0
26547.0
37206.0
42829.0
112107.0
118848.0
150236.0
45836.0
26484.0
73462.0
40916.0
214984.0
42829.0
195782.0
21645.0
26002.0
206809.0
68813.0
42829.0
181377.0
45836.0
112107.0
36846.0
26510.0
37206.0
214984.0
45836.0
26547.0
37206.0
5710.0
150236.0
37216.0
112107.0
42592.0
118848.0
150236.0
206809.0
42829.0
26484.0
214984.0
45836.0
40916.0
42592.0
21645.0
37206.0
195782.0
37216.0
26547.0
73462.0
26002.0
5710.0
112107.0
118848.0
26510.0
42829.0
206809.0
40916.0
26484.0
5710.0
37206.0
37216.0
150236.0
21645.0
118848.0
195782.0
45836.0
26547.0
112107.0
73462.0
214984.0
26002.0
37216.0
42829.0
5710.0
40916.0
26484.0
150236.0
206809.0
26510.0
37206.0
195782.0
214984.0
118848.0
26547.0
21645.0
209422.0
26002.0
40916.0
73462.0
5710.0
26484.0
278844.0
26510.0
45836.0
42829.0
112107.0
195782.0
214984.0
73462.0
21645

nan
nan
15361.0
32633.0
13436.0
13320.0
nan
nan
24341.0
43313.0
nan
nan
114822.0
13390.0
75579.0
nan
43313.0
24341.0
68740.0
13436.0
32633.0
nan
nan
13320.0
13454.0
114822.0
nan
75579.0
nan
nan
13436.0
13390.0
nan
nan
114817.0
24341.0
nan
13320.0
68740.0
32633.0
nan
nan
114822.0
114817.0
nan
nan
nan
13436.0
68740.0
114822.0
nan
136539.0
nan
13494.0
nan
24341.0
13436.0
69004.0
68835.0
13390.0
nan
13320.0
nan
24341.0
75579.0
nan
32633.0
136539.0
nan
nan
13390.0
nan
13454.0
nan
nan
24341.0
43313.0
69004.0
68835.0
nan
13471.0
13494.0
13436.0
nan
nan
136539.0
nan
nan
13390.0
nan
nan
43313.0
13471.0
68835.0
13436.0
nan
nan
13494.0
24341.0
114822.0
13471.0
nan
24341.0
43313.0
nan
13390.0
nan
nan
nan
114822.0
nan
13390.0
nan
nan
nan
75579.0
nan
nan
13473.0
43313.0
47249.0
32633.0
nan
43313.0
nan
32633.0
nan
13473.0
13454.0
68740.0
nan
75579.0
nan
nan
nan
13320.0
nan
68768.0
114822.0
206801.0
nan
13454.0
68768.0
nan
47249.0
68740.0
nan
13320.0
43313.0
75579.0
32633.0
nan
115006.0
114817.0
nan
7

19515.0
nan
nan
nan
nan
nan
nan
nan
150077.0
nan
nan
nan
nan
40963.0
40604.0
nan
97363.0
nan
nan
nan
33550.0
nan
115171.0
nan
43283.0
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
97365.0
nan
nan
150077.0
nan
33550.0
40963.0
30582.0
nan
163843.0
43283.0
nan
nan
22125.0
nan
nan
45245.0
nan
nan
97498.0
nan
33550.0
nan
nan
nan
nan
nan
nan
nan
nan
40963.0
30582.0
45245.0
nan
163843.0
43283.0
45306.0
nan
40604.0
97363.0
40610.0
150077.0
13345.0
97498.0
69341.0
33550.0
40963.0
40604.0
30582.0
19515.0
163843.0
43283.0
45306.0
nan
97363.0
nan
150275.0
150077.0
13345.0
97498.0
69341.0
33550.0
40963.0
40604.0
30967.0
19515.0
163843.0
43283.0
97498.0
nan
97363.0
nan
96844.0
150077.0
13345.0
45306.0
69341.0
33550.0
40963.0
40604.0
30967.0
19515.0
163843.0
43283.0
97498.0
33550.0
97363.0
nan
150275.0
150077.0
13345.0
45306.0
69341.0
nan
40963.0
40604.0
30967.0
97363.0
19515.0
163843.0
97498.0
33550.0
19434.0
163843.0
40610.0
43283.0
13345.0
45306.0
11716.0
nan
nan
nan
nan
nan
nan
97498.0


30989.0
280969.0
167285.0
19515.0
104986.0
361717.0
120621.0
194632.0
261348.0
206711.0
500575.0
13345.0
33431.0
164316.0
35453.0
40604.0
163887.0
186705.0
500575.0
167285.0
361717.0
19344.0
120621.0
124127.0
280969.0
206711.0
194632.0
361717.0
128083.0
261348.0
164316.0
35453.0
40604.0
19515.0
163887.0
186705.0
30989.0
500575.0
150066.0
33431.0
280969.0
167285.0
124127.0
40959.0
30989.0
426359.0
164316.0
97363.0
355347.0
30657.0
570468.0
186705.0
45306.0
40959.0
30989.0
120621.0
33431.0
30657.0
13345.0
52160.0
380949.0
164354.0
194632.0
96836.0
40604.0
426359.0
164316.0
361717.0
19515.0
181433.0
45306.0
40959.0
13345.0
571298.0
33431.0
30989.0
380949.0
30657.0
570468.0
164354.0
268306.0
164316.0
571298.0
361717.0
19515.0
96836.0
181433.0
40604.0
45306.0
40959.0
30989.0
426359.0
13345.0
380949.0
30657.0
181433.0
570468.0
164354.0
268306.0
19515.0
361717.0
164316.0
571298.0
355347.0
33431.0
40604.0
45306.0
40959.0
30989.0
30657.0
355347.0
426359.0
570468.0
181433.0
380949.0
164354.0
268

47559.0
186891.0
23114.0
33914.0
299380.0
148483.0
32562.0
13471.0
41066.0
23114.0
23141.0
14100.0
47559.0
186891.0
299380.0
23141.0
32658.0
115675.0
32562.0
13471.0
41066.0
23114.0
33914.0
299380.0
47559.0
13471.0
23141.0
354519.0
14100.0
115675.0
32562.0
35336.0
23114.0
33914.0
299380.0
14100.0
47559.0
13471.0
41066.0
109622.0
354519.0
115675.0
13471.0
23114.0
33914.0
109622.0
14100.0
148483.0
186891.0
23114.0
33914.0
299380.0
23141.0
14100.0
32562.0
186891.0
41066.0
33914.0
32658.0
115675.0
47559.0
252672.0
41066.0
241996.0
354519.0
14100.0
47559.0
32562.0
186891.0
23114.0
23141.0
115675.0
13471.0
41066.0
33914.0
354519.0
14100.0
115675.0
47559.0
32562.0
186891.0
34213.0
299380.0
23141.0
186891.0
41066.0
33914.0
32658.0
14100.0
115675.0
47559.0
32562.0
252672.0
34213.0
241996.0
23141.0
32562.0
13471.0
241996.0
32658.0
14100.0
115675.0
32418.0
186891.0
41066.0
34213.0
33914.0
109622.0
32418.0
32562.0
35336.0
41066.0
299380.0
354519.0
32562.0
186891.0
241996.0
109622.0
32658.0
14100.0

37456.0
74736.0
104986.0
37579.0
75195.0
37421.0
170323.0
41466.0
71498.0
109222.0
28955.0
32657.0
37456.0
74736.0
33845.0
113695.0
37439.0
33753.0
46531.0
33764.0
30657.0
37623.0
37421.0
74736.0
170323.0
41466.0
71498.0
96613.0
37579.0
32657.0
104986.0
75195.0
113695.0
37439.0
33753.0
46531.0
33764.0
30657.0
28955.0
37623.0
37456.0
33845.0
96623.0
37579.0
71498.0
96613.0
32657.0
104986.0
37439.0
75195.0
37421.0
74736.0
28955.0
37623.0
37456.0
41466.0
33845.0
113695.0
33753.0
46531.0
33764.0
30657.0
75195.0
37579.0
74736.0
170323.0
71498.0
96613.0
33764.0
32657.0
104986.0
37439.0
33845.0
33753.0
30657.0
196126.0
113695.0
46531.0
28955.0
41466.0
37421.0
37579.0
32657.0
104986.0
37439.0
75195.0
74736.0
170323.0
37623.0
71498.0
96613.0
33764.0
109222.0
170323.0
75195.0
32657.0
71498.0
74736.0
37439.0
33764.0
37623.0
33845.0
163843.0
170323.0
71498.0
37623.0
33764.0
109222.0
37439.0
33732.0
33845.0
75195.0
196126.0
37579.0
28955.0
30657.0
104986.0
46531.0
33753.0
37421.0
40672.0
41466.0
37

37421.0
300532.0
71498.0
246201.0
37656.0
177126.0
25563.0
33764.0
104986.0
37579.0
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
95231.0
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
95231.0
nan
39420.0
42276.0
25811.0
41618.0
42247.0
42276.0
67311.0
25947.0
95231.0
nan
39420.0
36479.0
25811.0
95231.0
34485.0
42247.0
42276.0
25947.0
41601.0
36479.0
39420.0
67311.0
25811.0
34485.0
42276.0
41601.0
95231.0
42247.0
39420.0
36479.0
25811.0
67311.0
25947.0
95231.0
34485.0
42276.0
42247.0
25947.0
nan
nan
nan
nan
nan
39420.0
41601.0
36479.0
66902.0
25811.0
34485.0
42276.0
42247.0
66902.0
25947.0
95231.0
36479.0
25811.0
22700.0
nan
34485.0
36479.0
42247.0
66902.0
25947.0
22700.0
25811.0
95231.0
42276.0
nan
22700.0
25760.0
nan
66902.0
25947.0
95231.0
34485.0
67374.0
41601.0
31470.0
nan
nan
nan
nan
nan
nan
nan
nan


## Partie 1 - Description du problème

En ce qui concernce le football, le pari sportif est basé sur l'ignorance du resultat d'un match opposant deux equipe. <br> Reussir a predire qu'un match se termine par une victoire ou une defaite d'une equipe ou un nul des deux serait le gral du parie.<br> Dans de ce notebook, nous allons creer un classifieur qui nous permettra de predire les resultats de matchs avec un certains taux de succes, et ceux en nous basant sur les données collectés sur des equipes de 11 leagues europeenes et sur plusieurs saisons.



## Partie 2 - Modéle

### Bagging de perceptron 

* Creation du LabeledSet

In [39]:
class LabeledSet:  
    
    def __init__(self, input_dimension):
        self.input_dimension = input_dimension
        self.nb_examples = 0
    
    def addExample(self,vector,label):
        if (self.nb_examples == 0):
            self.x = np.array([vector])
            self.y = np.array([label])
        else:
            self.x = np.vstack((self.x, vector))
            self.y = np.vstack((self.y, label))
        
        self.nb_examples = self.nb_examples + 1
    
    #Renvoie la dimension de l'espace d'entrée
    def getInputDimension(self):
        return self.input_dimension
    
    #Renvoie le nombre d'exemples dans le set
    def size(self):
        return self.nb_examples
    
    #Renvoie la valeur de x_i
    def getX(self, i):
        return self.x[i]
        
    
    #Renvouie la valeur de y_i
    def getY(self, i):
        return(self.y[i])

Modéle

In [40]:
import random
import numpy as np
import pandas as pd
import matplotlib as plt
import math
import graphviz
import datetime

In [41]:
class Classifier:
    def __init__(self,input_dimension):
        """ Constructeur """
        raise NotImplementedError("Please Implement this method")
    
    
    # Permet de calculer la prediction sur x => renvoie un score
    def predict(self,x):
        raise NotImplementedError("Please Implement this method")

    
    # Permet d'entrainer le modele sur un ensemble de données étiquetés
    def train(self,labeledSet):
        raise NotImplementedError("Please Implement this method")
    
    # Permet de calculer le taux de bonne classification
    def accuracy(self,set):
        nb_ok=0
        for i in range(set.size()):
            score = self.predict(set.getX(i))
            if (score*set.getY(i)>0):
                nb_ok = nb_ok+1
        acc = nb_ok/(set.size() * 1.0)
        return acc


In [42]:
#np.random.rand(input_dimension)

In [43]:
class Perceptron(Classifier):
    
    def __init__(self, input_dimension, learning_rate, nombre_iterations, debut_saison):
        self.input_dimension = input_dimension
        self.learning_rate = learning_rate
        self.nombre_iterations = nombre_iterations
        self.w = np.array([ 1.0,  1.0])
        self.debut_saison = debut_saison
        self.seuil = random.uniform(0,1)
        print(self.seuil)
        #self.proba_domicile = proba_domicile
    
    def predict(self, x): #renvoi +1 si l'equipe a domicile gagne
        noteD=x[0]
        noteE=x[1]
        print("yé suis le perceptron regular")
        if (abs(noteD-noteE)<self.seuil/2.0):
            #print(str(noteD)+" dom et ext "+str(noteE)+" et seuil "+str(self.seuil/2))
            return 0
        #calculer le produit scalaire que dans le cas ou la diff de notes est grande
        z = np.dot((noteD,noteE), self.w) #a confirmer sur la logique du calcul
        if z > 0:
            print("return 1")
            return +1 
        else:
            print("return -1")
            return -1
        
    def train(self, labeledSet):
        datas = labeledSet
        t = labeledSet.size()
        app = LabeledSet(self.input_dimension)
        test = LabeledSet(self.input_dimension)
        #noteD=LabeledSet.getX(i)[0]
        #noteE=LabeledSet.getX(i)[1]
        for i in range(t):
            if i < t*0.8:
                app.addExample(datas.getX(i), datas.getY(i))
            else :
                test.addExample(datas.getX(i), datas.getY(i))
        for j in range (self.nombre_iterations): #ou self.nombre_iterations
            for i in range(t):
                #calc = self.predict(datas.getX(i)) * datas.getY(i)
                noteD=datas.getX(i)[0]
                noteE=datas.getX(i)[1]
                if (self.predict(datas.getX(i)) != datas.getY(i)):
                    if (self.predict(datas.getX(i)) == 0):
                        self.seuil -= 0.1
                        print("diminuer au seuil")
                    elif (datas.getY(i) == 0):
                        self.seuil += 0.1
                        print("ajouter au seuil")
                #if calc < 0:
                    #maj de w += epsilone * x * y
                    else: #des notes ou getX????????????************************************
                        print("reajuster w")
                        self.w[0] = self.w[0] + self.learning_rate * noteD * datas.getY(i)
                        self.w[1] = self.w[1] + self.learning_rate * noteE * datas.getY(i)
            #if j % 10 == 0:
            #    self.accuracy(test)

    def accuracy(self, dataSet):
        juste = 0
        for i in range(dataSet.size()):
            print(str(self.predict(dataSet.getX(i)))+" trouvé et attendu "+str(dataSet.getY(i)))
            if self.predict(dataSet.getX(i)) == dataSet.getY(i):
                #print(str(self.predict(dataSet.getX(i)))+" trouvé et attendu "+str(dataSet.getY(i)))
                juste += 1
        return float(juste)/float(dataSet.size())
            #split

In [44]:
class KernelBias:
    def transform(self,x):
        return np.concatenate((x, [1]))

In [45]:
class PerceptronKernel:
    def __init__(self,dimension_kernel,learning_rate,nombre_iterations,kernel,debut_saison):
        self.nombre_iterations = nombre_iterations
        self.learning_rate = learning_rate
        self.dimension_kernel = dimension_kernel
        self.w = np.random.normal(size=dimension_kernel)
        self.kernel = kernel
        self.debut_saison = debut_saison
        self.seuil = random.uniform(0,1)
        
    #Permet de calculer la prediction sur x => renvoie un score
    def predict(self,x):
        
        noteD=x[0]+0.5
        noteE=x[1]
        if (abs(noteD-noteE)<self.seuil/2.0):
            print(str(noteD)+" dom et ext "+str(noteE)+" et seuil "+str(self.seuil/2))
            return 0
        #calculer le produit scalaire que dans le cas ou la diff de notes est grande
        z = np.dot(self.kernel.transform(x), self.w) #a confirmer sur la logique du calcul
        if z > 0:
            print("return 1")
            return +1 
        else:
            print("return -1")
            return -1
        
        
       
    
    #Permet d'entrainer le modele sur un ensemble de données
    def train(self,labeledSet):
        
        datas = labeledSet
        t = labeledSet.size()
        app = LabeledSet(datas.getInputDimension())
        test = LabeledSet(datas.getInputDimension())
        #noteD=LabeledSet.getX(i)[0]
        #noteE=LabeledSet.getX(i)[1]
        for i in range(t):
            if i < t*0.8:
                app.addExample(datas.getX(i), datas.getY(i))
            else :
                test.addExample(datas.getX(i), datas.getY(i))
        for j in range (self.nombre_iterations): #ou self.nombre_iterations
            for i in range(t):
                #calc = self.predict(datas.getX(i)) * datas.getY(i)
                noteD=datas.getX(i)[0]
                noteE=datas.getX(i)[1]
                if (self.predict(datas.getX(i)) != datas.getY(i)):
                    if (self.predict(datas.getX(i)) == 0):
                        self.seuil -= 0.2
                        print("diminuer au seuil")
                    elif (datas.getY(i) == 0):
                        self.seuil += 0.2
                        print("ajouter au seuil")
                #if calc < 0:
                    #maj de w += epsilone * x * y
                    else: #des notes ou getX????????????************************************
                        print("reajuster w")
                        self.w += self.learning_rate * datas.getY(j) * self.kernel.transform(datas.getX(j))
                            
        
        
        
       
    
    #Permet de calculer la qualité du système 
    def accuracy(self,dataSet):
        nb_ok = 0
        for i in range(dataSet.size()):
            output = self.predict(dataSet.getX(i))
            if (output * dataSet.getY(i) > 0):
                nb_ok = nb_ok + 1
        acc = nb_ok / (dataSet.size() * 1.0)
        return acc
        

In [46]:
class ClassifierBaggingKernelPerceptron(Classifier):
    def __init__(self, nbPerc, pourcentage, input_dimension, learning_rate, nombre_iterations, boole,debut_saison,dimension_kernel,kernel):
        self.nbPerc=nbPerc
        self.pourcentage =pourcentage
        self.input_dimension = input_dimension
        self.learning_rate = learning_rate
        self.nombre_iterations = nombre_iterations
        self.boole = boole
        self.debut_saison=debut_saison        
        self.ensPerc=[]
        self.dimension_kernel = dimension_kernel
        self.w = np.random.normal(size=dimension_kernel)
        self.kernel = kernel
        #self.proba_domicile = proba_domicile
    
    def train(self,set):
        #x est l'id d'une equipe interieure, y exterieure
        for i in range(self.nbPerc):
            percep = PerceptronKernel(self.dimension_kernel,self.learning_rate,self.nombre_iterations,self.kernel,self.debut_saison)
            m=int(set.size()*self.pourcentage)
            echantillon=echantillonLS(set,m,self.boole)
            percep.train(echantillon)
            self.ensPerc.append(percep)
    
    
    def predict(self,x):
        moins,plus,zero=0,0,0
        l=[0,0,0] # une liste qui represente les labels predits l[0]=>-1; l[1]=>0; l[2]=>+1
        for i in self.ensPerc:
            res=i.predict(x)
            
            if (res == 0):
                l[1] += 1
                return 0
            elif (res==1):
                l[2] += 1
                return 1
            else:
                l[0] += 1
                return -1
        return (l[np.argmax(l)]-1)
    
    def accuracy(self,set):
        nb_ok=0
        for i in range(set.size()):
            print(str(self.predict(set.getX(i)))+" trouvé et attendu "+str(set.getY(i)))
            score = self.predict(set.getX(i))
            if (score==set.getY(i)):
                nb_ok = nb_ok+1
        acc = nb_ok/(set.size() * 1.0)
        return acc

In [47]:
def tirage(X, m, r) :
    if r :
        return[X[random.choice(X)] for i in range(0, m)]
    else :
        return random.sample(X, m)

In [48]:
def echantillonLS(LSet,m,r):
    v=[i for i in range(LSet.size())]
    v=tirage(v,m,r)
    dim=LSet.getInputDimension()
    X=LabeledSet(dim)
    for j in v:
        X.addExample(LSet.getX(j),LSet.getY(j))
    
    return X

In [49]:
class ClassifierBaggingPerceptron(Classifier):
    def __init__(self, nbPerc, pourcentage, input_dimension, learning_rate, nombre_iterations, boole,debut_saison):
        self.nbPerc=nbPerc
        self.pourcentage =pourcentage
        self.input_dimension = input_dimension
        self.learning_rate = learning_rate
        self.nombre_iterations = nombre_iterations
        self.boole = boole
        self.debut_saison=debut_saison        
        self.ensPerc=[]
        #self.proba_domicile = proba_domicile
    
    def train(self,set):
        #x est l'id d'une equipe interieure, y exterieure
        for i in range(self.nbPerc):
            percep = Perceptron(self.input_dimension, self.learning_rate, self.nombre_iterations, self.debut_saison)
            m=int(set.size()*self.pourcentage)
            echantillon=echantillonLS(set,m,self.boole)
            percep.train(echantillon)
            self.ensPerc.append(percep)
    
    
    def predict(self,x):
        moins,plus,zero=0,0,0
        l=[0,0,0] # une liste qui represente les labels predits l[0]=>-1; l[1]=>0; l[2]=>+1
        for i in self.ensPerc:
            res=i.predict(x)
            
            if (res == 0):
                l[1] += 1
                return 0
            elif (res==1):
                l[2] += 1
                return 1
            else:
                l[0] += 1
                return -1
        return (l[np.argmax(l)]-1)
    
    def accuracy(self,set):
        nb_ok=0
        for i in range(set.size()):
            print(str(self.predict(set.getX(i)))+" trouvé et attendu "+str(set.getY(i)))
            score = self.predict(set.getX(i))
            if (score==set.getY(i)):
                nb_ok = nb_ok+1
        acc = nb_ok/(set.size() * 1.0)
        return acc

## Partie 3 - Code

** preparation des données **

pour chaque equipe on fait sa note globale qui est la moyenne des notes des joueurs de l'equipe

In [50]:
saison_av = "2009/2010"
saison_ap = "2010/2011"
league=1

In [51]:
def matchsaison(saison,numleague):
    bdtot= LabeledSet(2)
    for i in range(len(match)):
        
        
        if(match.iloc[i]["league_id"]==numleague and saison==str(match.iloc[i]["season"])):
            res=match.iloc[i]["home_team_goal"]-match.iloc[i]["away_team_goal"]
            lab=0
            if (res>0):
                lab=+1
            elif(res<0):
                lab=-1
            else:
                lab=0
            bdtot.addExample((match.iloc[i]['home_team_api_id'],match.iloc[i]['away_team_api_id']),lab)
            #print(match.iloc[i]['home_team_api_id'],match.iloc[i]['away_team_api_id'],lab)
    return bdtot        

In [52]:
lab=matchsaison(saison_av,league)
labT=matchsaison(saison_ap,league)
print(lab.size())

210


In [53]:
def MatchesSaisonLeague(championnat,saison):
    
   
    #avoir les matches joué a domicile pendant la saison en question
    gb=match.groupby("league_id")
    matchesSaisonLeague=gb.get_group(championnat)
    matchesSaisonLeague=matchesSaisonLeague[matchesSaisonLeague.season == saison]
    matchesSaisonLeague=matchesSaisonLeague[["home_team_api_id","away_team_api_id","date",'home_player_1', 'home_player_2', 'home_player_3', "home_player_4", "home_player_5",
               "home_player_6", "home_player_7", "home_player_8", "home_player_9", "home_player_10",
               "home_player_11", "away_player_1", "away_player_2", "away_player_3", "away_player_4",
               "away_player_5", "away_player_6", "away_player_7", "away_player_8", "away_player_9",
               "away_player_10", "away_player_11"]]
    
    return matchesSaisonLeague

In [54]:
def StatJoueurSaison(saison):
    
    playerattributes['date']  = pd.to_datetime(playerattributes['date'])
    df_filtre = playerattributes[(playerattributes.date > datetime.date(int(saison[0:4]),9,1)) & (playerattributes.date < datetime.date(int(saison[5:]),8,30))]
    df_filtre=df_filtre[['player_api_id','overall_rating']]
    
    return df_filtre

In [55]:
def StatEquipesLeagueSaison(saison,championnat):
    matchesSaisonLeague=MatchesSaisonLeague(championnat,saison)
    statJoueur=StatJoueurSaison(saison)
    statEquipes=pd.DataFrame(({'team_api_id': [], 'overall_rating': []}))
    ColonnesJoueurs_home = ['home_player_1', 'home_player_2', 'home_player_3', "home_player_4", "home_player_5",
                           "home_player_6", "home_player_7", "home_player_8", "home_player_9", "home_player_10",
                           "home_player_11"]


    #Pour recuperer les joueurs qui font partie des 11 titulaires des matches a l'exterieur
    ColonnesJoueurs_away=["away_player_1", "away_player_2", "away_player_3", "away_player_4","away_player_5",
                          "away_player_6", "away_player_7","away_player_8", "away_player_9","away_player_10",
                          "away_player_11"]


    #avoir les matches joué a domicile pendant la saison en question
    
   


    #avoir les matches joué a l'exterieur pendant la saison en question
    
    
    ligne=0
    
    for matche in matchesSaisonLeague.itertuples():
        #print(type(matche))
        #sam = statEquipes.loc[statEquipes["team_api_id"] == matche["home_team_api_id"]]
        #lig, col =sam.shape
        #print(lig)
        if matche.home_team_api_id in list(statEquipes.team_api_id):
            continue
        else:
            
            gb=matchesSaisonLeague.groupby("home_team_api_id")
            myteam_home=gb.get_group(matche.home_team_api_id)
            
           
            gb2=matchesSaisonLeague.groupby("away_team_api_id")
            myteam_away=gb.get_group(matche.home_team_api_id)
            NotesJoueursD=pd.DataFrame()
            for postejoueur in ColonnesJoueurs_home:
                # avoir la liste des joueurs ayant joué dans ce poste

                effectifs=myteam_home[postejoueur].dropna() # on utilise dropna pour eliminer tous les nan
                #print(effectifs)
                NotesJoueursD[postejoueur]=[0]
                noteglob=pd.DataFrame()
                note=list()
                for i in range(len(effectifs)):
                    noteglob=pd.DataFrame()

                    df_filtre=statJoueur[['player_api_id','overall_rating']]
                    note.append(df_filtre[df_filtre['player_api_id']==effectifs.iloc[i]]['overall_rating'].mean())              
                    #print(df_masked[df_masked['player_api_id']==effectifs.iloc[i]])
                    #NotesJoueurs[postejoueur]=df_masked[df_masked['player_api_id']==effectifs.iloc[i]]['overall_rating']

                noteglob = pd.DataFrame(np.array(note), columns = ["a"]).dropna()

                NotesJoueursD[postejoueur]=[noteglob.mean()[0]]
            #print(NotesJoueursD)

            NotesJoueursA=pd.DataFrame()
            for postejoueur in ColonnesJoueurs_away:
                # avoir la liste des joueurs ayant joué dans ce poste

                effectifs=myteam_away[postejoueur].dropna() # on utilise dropna pour eliminer tous les nan

                NotesJoueursA[postejoueur]=[0]
                noteglob=pd.DataFrame()
                note=list()
                for i in range(len(effectifs)):
                    noteglob=pd.DataFrame()

                    note.append(df_filtre[df_filtre['player_api_id']==effectifs.iloc[i]]['overall_rating'].mean())              
                    #print(df_masked[df_masked['player_api_id']==effectifs.iloc[i]])
                    #NotesJoueurs[postejoueur]=df_masked[df_masked['player_api_id']==effectifs.iloc[i]]['overall_rating']
                noteglob = pd.DataFrame(np.array(note), columns = ["a"]).dropna()

                NotesJoueursA[postejoueur]=[noteglob.mean()[0]]
            #print(NotesJoueursA.mean(axis = 1))
        statEquipes.loc[ligne]=[matche.home_team_api_id, ((NotesJoueursA.mean(axis = 1)+NotesJoueursD.mean(axis = 1))/2)[0]]
        ligne+=1
    
    
    statEquipes.rename(columns={'overall_rating': 'team_api_id', 'team_api_id': 'overall_rating'}, inplace=True)
   
    return statEquipes.groupby('team_api_id', as_index=False).max()
                        
    
    

In [ ]:
stats_equipes_league_saison=StatEquipesLeagueSaison(saison_av,1)

In [ ]:
nbmatches=math.factorial(18)/math.factorial(16)
print("il y a 18 equipes dans le championnat belge durant la saison 2008/2009, donc le nombre de matches disputés est: ",nbmatches)

In [ ]:
match.iloc(0)
str(match["season"][0])

In [ ]:
playerattributes

In [ ]:
def noteEquipe(idequipe,saison):
    #Pour recuperer les joueurs qui font partie des 11 titulaires des matches a domicile
    ColonnesJoueurs_home = ['home_player_1', 'home_player_2', 'home_player_3', "home_player_4", "home_player_5",
           "home_player_6", "home_player_7", "home_player_8", "home_player_9", "home_player_10","home_player_11"]


    #Pour recuperer les joueurs qui font partie des 11 titulaires des matches a l'exterieur
    ColonnesJoueurs_away=["away_player_1", "away_player_2", "away_player_3", "away_player_4","away_player_5", "away_player_6", "away_player_7",
                          "away_player_8", "away_player_9","away_player_10", "away_player_11"]


    #avoir les matches joué a domicile pendant la saison en question
    gb=match.groupby("home_team_api_id")
    myteam_home=gb.get_group(idequipe)
    myteam_home=myteam_home[myteam_home.season == saison]


    #avoir les matches joué a l'exterieur pendant la saison en question
    gb2=match.groupby("away_team_api_id")
    myteam_away=gb.get_group(idequipe)
    myteam_away[myteam_away.season == saison]
    playerattributes['date']  = pd.to_datetime(playerattributes['date'])
    df_filtre = playerattributes[(playerattributes.date > datetime.date(int(saison[0:4]),9,1)) & (playerattributes.date < datetime.date(int(saison[5:]),8,30))]
    df_filtre=df_filtre[['player_api_id','overall_rating']]

    
    
    NotesJoueursD=pd.DataFrame()
    for postejoueur in ColonnesJoueurs_home:
        # avoir la liste des joueurs ayant joué dans ce poste

        effectifs=myteam_home[postejoueur].dropna() # on utilise dropna pour eliminer tous les nan
        #print(effectifs)
        NotesJoueursD[postejoueur]=[0]
        noteglob=pd.DataFrame()
        note=list()
        for i in range(len(effectifs)):
            noteglob=pd.DataFrame()

            df_filtre=df_filtre[['player_api_id','overall_rating']]
            note.append(df_filtre[df_filtre['player_api_id']==effectifs.iloc[i]]['overall_rating'].mean())              
            #print(df_masked[df_masked['player_api_id']==effectifs.iloc[i]])
            #NotesJoueurs[postejoueur]=df_masked[df_masked['player_api_id']==effectifs.iloc[i]]['overall_rating']

        noteglob = pd.DataFrame(np.array(note), columns = ["a"]).dropna()

        NotesJoueursD[postejoueur]=[noteglob.mean()[0]]
    #print(NotesJoueursD)

    NotesJoueursA=pd.DataFrame()
    for postejoueur in ColonnesJoueurs_away:
        # avoir la liste des joueurs ayant joué dans ce poste

        effectifs=myteam_away[postejoueur].dropna() # on utilise dropna pour eliminer tous les nan

        NotesJoueursA[postejoueur]=[0]
        noteglob=pd.DataFrame()
        note=list()
        for i in range(len(effectifs)):
            noteglob=pd.DataFrame()

            note.append(df_filtre[df_filtre['player_api_id']==effectifs.iloc[i]]['overall_rating'].mean())              
            #print(df_masked[df_masked['player_api_id']==effectifs.iloc[i]])
            #NotesJoueurs[postejoueur]=df_masked[df_masked['player_api_id']==effectifs.iloc[i]]['overall_rating']
        noteglob = pd.DataFrame(np.array(note), columns = ["a"]).dropna()

        NotesJoueursA[postejoueur]=[noteglob.mean()[0]]
    #print(NotesJoueursA.mean(axis = 1))
    return ((NotesJoueursA.mean(axis = 1)+NotesJoueursD.mean(axis = 1))/2)[0]


In [ ]:

float(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==9987.0]["overall_rating"])

In [ ]:
noteEquipe(9987,saison_av)


In [ ]:
playerattributes[playerattributes['player_api_id']==170323]

In [ ]:
import datetime
#datetime.datetime.strptime(playerattributes.date[0], '%Y-%m-%d %H:%M:%S')

In [ ]:
stats_equipes_league_saison

## Partie 4 - Protocole expérimental

In [ ]:
input_dimension = 2
nbPerc = 11
pourcentage = 0.3
input_dimension = 2
learning_rate = 0.25
nb_ite = 51
boole = True  


* **taux de reussite 40% : <br> **
input_dimension = 2  nbPerc = 7  pourcentage = 0.3  input_dimension = 2   learning_rate = 0.25   nb_ite = 35     boole = True  

In [ ]:
app=LabeledSet(input_dimension)
test=LabeledSet(input_dimension)
for i in range(lab.size()):  
    
    if(len(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==lab.getX(i)[0]]["overall_rating"])>0):
        NoteD=float(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==lab.getX(i)[0]]["overall_rating"])
    else:
        NoteD=noteEquipe(lab.getX(i)[0],saison_ap)
    
    if(len(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==lab.getX(i)[1]]["overall_rating"])>0):
        NoteE=float(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==lab.getX(i)[1]]["overall_rating"])
    else:        
        NoteE=noteEquipe(lab.getX(i)[1],saison_ap)    
    
    
    if(len(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==labT.getX(i)[0]]["overall_rating"])>0):
        NoteDT=float(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==labT.getX(i)[0]]["overall_rating"])
    else:
        NoteDT=noteEquipe(labT.getX(i)[0],saison_ap)
    
    if(len(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==labT.getX(i)[1]]["overall_rating"])>0):
        NoteET=float(stats_equipes_league_saison[stats_equipes_league_saison["team_api_id"]==labT.getX(i)[1]]["overall_rating"])
    
    else:        
        NoteET=noteEquipe(labT.getX(i)[1],saison_ap)  
    
    
    #print(NoteD,NoteE)
    #print(NoteDT,NoteET)
    app.addExample([NoteD,NoteE],lab.getY(i))
    test.addExample([NoteDT,NoteET],labT.getY(i))

    

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline 

def plot2DSet(set):
    """ LabeledSet -> NoneType
        Hypothèse: set est de dimension 2
        affiche une représentation graphique du LabeledSet
        remarque: l'ordre des labels dans set peut être quelconque
    """
    S_pos = set.x[np.where(set.y == 1),:][0]      # tous les exemples de label +1
    S_neg = set.x[np.where(set.y == -1),:][0]     # tous les exemples de label -1
    S_nul = set.x[np.where(set.y == 0),:][0]     # tous les exemples de label 0
    plt.scatter(S_pos[:,0],S_pos[:,1],marker='o')
    plt.scatter(S_neg[:,0],S_neg[:,1],marker='x')
    plt.scatter(S_nul[:,0],S_nul[:,1],marker='v')


In [ ]:
plot2DSet(app)

In [ ]:
percep = ClassifierBaggingPerceptron(nbPerc, pourcentage, input_dimension, learning_rate, nb_ite, boole,2009)
percep.train(app)
print("accuracy apprentissage:", percep.accuracy(app)) 

## Partie 5 - Résultats

In [ ]:
print("accuracy teste:", percep.accuracy(test)) 









bizarité :=> bagging a ça propre methode de predict, Par contre lors du test de l'accuracy, on utilise la methode accuracy du perceptron et celle ci appelle la methode predict de perceptron et non bagging !

In [ ]:

percepKern = ClassifierBaggingKernelPerceptron(nbPerc, pourcentage, input_dimension, learning_rate, nb_ite, boole,2009,3,KernelBias())
percepKern.train(app)
print("accuracy apprentissage:", percepKern.accuracy(app)) 

In [ ]:
print("accuracy apprentissage:", percepKern.accuracy(test)) 

In [ ]:
def plot_frontiere(set,classifier,step=20):
    """ LabeledSet * Classifier * int -> NoneType
        Remarque: le 3e argument est optionnel et donne la "résolution" du tracé
        affiche la frontière de décision associée au classifieur
    """
    mmax = set.x.max(0)
    mmin = set.x.min(0)
    x1grid, x2grid = np.meshgrid(np.linspace(mmin[0], mmax[0], step), np.linspace(mmin[1], mmax[1], step))
    grid = np.hstack((x1grid.reshape(x1grid.size, 1), x2grid.reshape(x2grid.size, 1)))
    
    # calcul de la prediction pour chaque point de la grille
    res = np.array([classifier.predict(grid[i,:]) for i in range(len(grid)) ])
    res = res.reshape(x1grid.shape)
    # tracé des frontieres
    plt.contourf(x1grid, x2grid, res, colors=["red", "cyan"], levels=[-1000,0,1000], linewidth=2) 
    

In [ ]:
plot_frontiere(test,percepKern,20)
plot2DSet(test)

## Partie 6 - Analyse

Meilleure score est de 56% de réussite.
On a dut modifier notre code pour le perceptron et le bagging de perceptrons, dans le but de renvoyé +1 pour une prédiction de victoire pour l'equipe a domicile, -1 une perte et 0 un nul. On a rajouté un seuil, qu'on utilise dans le cas ou la difference de notes de 2 équipes est inférieure a a ce seuil la donc on prédit un match nul, si on se trompe lors de cette prédiction alor le seuil est décrementé, et reciproquement si l'ont ne prédit pas une égalité or qu'il fallais prédire l'égalité alors le seuil est incrémenté. D'autres